Premake generator: better integration

perseoGI opened the pull request:
Changelog: Feature: Add premake toolchain and improved premake integration in conan with new premake5.
Docs: https://github.com/conan-io/docs/pull/4090

Follow https://github.com/premake/premake-core/discussions/2441
And test project: https://github.com/perseoGI/conan-premake-example

perseoGI committed:
WIP: added premake toolchain and updated premake generator

memsharded commented on conan/tools/premake/toolchain.py:
```diff
@@ -0,0 +1,71 @@
+    _premake_file_template = textwrap.dedent(
+        """\
+    #!lua
+    include("conandeps.premake5.lua")
```
What if the current package has no dependencies and the PremakeDeps is not used and this file doesn't exist?

perseoGI replied:
The `include` of the `conandeps.premake5.lua` file is controlled under a "file existence flag".
It will work as expected! It won't be included as it is not needed and won't affect the premake5 invocation.

perseoGI committed:
Updated premake toolchain

perseoGI committed:
Created build method on Premake, fix cross building and map architectures

perseoGI committed:
Added extra flags on premake generator

perseoGI committed:
Stable version, not complete

perseoGI committed:
Detailed tooling

perseoGI committed:
Improved code generation identation

perseoGI committed:
Add option to disable projects

memsharded commented on conan/tools/premake/toolchain.py:
```diff
@@ -0,0 +1,296 @@
+from conan.tools.files import save
+from conan.tools.premake.premakedeps import PREMAKE_ROOT_FILE
+
+jinja_env = Environment(trim_blocks=True, lstrip_blocks=True)
```
No need to have this global

memsharded commented on conan/tools/premake/premake.py:
```diff
@@ -1,40 +1,96 @@
+
+        arch = str(self._conanfile.settings.arch)
+        if arch not in CONAN_TO_PREMAKE_ARCH:
+            raise ConanInvalidConfiguration(
```
ConanInvalidConfiguration is not intended for this point, only for ``validate()``

memsharded commented on conan/tools/premake/premake.py:
```diff
@@ -1,40 +1,96 @@
-        self.luafile = 'premake5.lua'   # Path to the root (premake5) lua file
+        # Path to the root (premake5) lua file
+        self.luafile = (Path(self._conanfile.source_folder) / "premake5.lua").as_posix()
+        print("luafile", self.luafile)
```
Convert print to `self.output``

memsharded commented on conan/tools/premake/premake.py:
```diff
@@ -1,40 +1,96 @@
         if "msvc" in self._conanfile.settings.compiler:
-            msvc_version = PREMAKE_VS_VERSION.get(str(self._conanfile.settings.compiler.version))
-            self.action = f'vs{msvc_version}'
+            msvc_version = PREMAKE_VS_VERSION.get(
+                str(self._conanfile.settings.compiler.version)
+            )
+            self.action = f"vs{msvc_version}"
```
Avoid formatting, it adds extra overhead for the review process.

memsharded commented on conan/tools/premake/premake.py:
```diff
@@ -1,40 +1,96 @@
+            self.action = f"vs{msvc_version}"
         else:
-            self.action = 'gmake2'
+            self.action = "gmake"
```
Why? 
Seems deprecated, and the recommended one is ``gmake2``? 
https://github.com/premake/premake-core/wiki/Using-Premake/9507d6baa7b52ac97f3eaa7eddaa5f6bb984fa39

perseoGI replied:
It used to be `gmake2` the new one, but the attached documentation is already deprecated.
This is the newest documentation: https://premake.github.io/docs/Using-Premake
`gmake` is the newest one and now, the legacy generator is called `gmakelegacy`.

See `$ premake --help` output:
``` 
...
ACTIONS

 clean             Remove all binaries and generated files
 codelite          Generate CodeLite project files
 gmake             Generate GNU makefiles for POSIX, MinGW, and Cygwin
 gmakelegacy       Generate GNU makefiles for POSIX, MinGW, and Cygwin
...
```

memsharded commented on conan/tools/premake/premake.py:
```diff
@@ -1,40 +1,96 @@
-    '192': '2019',
-    '193': '2022',
-    '194': '2022',  # still 2022
+    "190": "2015",
```
Avoid extra unnecessary formatting.

memsharded commented on conan/tools/premake/toolchain.py:
```diff
@@ -0,0 +1,296 @@
+
+    -- Generate workspace configurations
+    {% for workspace in workspaces.values() %}
+    {{ workspace._generate() }}
```
Private call, doesn't look great

memsharded commented on conan/tools/premake/toolchain.py:
```diff
@@ -0,0 +1,296 @@
+
+    def __init__(self, conanfile):
+        self._conanfile = conanfile
+        self._workspaces = {}
```
It is not very clear why the organization of the toolchain in workspaces.
It seems workspaces would be a feature of the project not a feature for the toolchain, but I have little idea about premake, so probably I need to undersand better about the workspaces in premake

perseoGI committed:
Removed global projects as it is not an stable solution

perseoGI committed:
Use ConanException instead

perseoGI committed:
Removed possibility of managing different workspaces

perseoGI committed:
Restore multi workspace allowing PremakeToolchain to be used as a plain generator

perseoGI committed:
Added more premake tests

memsharded commented on conan/tools/premake/constants.py:
```diff
@@ -0,0 +1,34 @@
+    "asm.js": "wasm32",
+}
+
+INDENT_LEVEL = 4
```
No need to make it a constant in another file, just make it a local constant in the place it is used

memsharded commented on conan/tools/premake/premake.py:
```diff
@@ -1,40 +1,91 @@
+        else:
+            build_type = str(self._conanfile.settings.build_type)
+            targets = "all" if targets is None else " ".join(targets)
+            self._conanfile.run(f"make config={build_type.lower()} {targets} -j")
```
Do not hardcode ``-j`` (I assume that will use the maximum cores?); but use the ``conf`` that models the cores for parallel building

perseoGI replied:
Good idea, I have just updated it

perseoGI committed:
Parametrize premake test

perseoGI committed:
Added some comments

perseoGI committed:
Removed old premake functional test

perseoGI committed:
Added docstring and a way of controlling verbosity on premake

perseoGI committed:
Fixed tests

perseoGI committed:
Revert extra formatting

perseoGI committed:
Remove print

memsharded commented on conan/tools/premake/premakedeps.py:
```diff
@@ -145,10 +146,11 @@ def generate(self):
+        #          ("Platform", self.architecture)]
+        # name = "".join("_%s" % v for _, v in props)
+        # return name.lower()
+        return f"{self.configuration}_{CONAN_TO_PREMAKE_ARCH[str(self.architecture)]}".lower()
```
Why changing the filenames of the "Deps" files? 
This used to generate different files for different architectures, like the dependencies for ``arch=armv6`` and ``arch=armv7`` would generate different files. Now, they will generate the same file, overwriting each other, why?

memsharded commented on conan/tools/premake/premakedeps.py:
```diff
@@ -4,9 +4,10 @@
 
 # Filename format strings
-PREMAKE_VAR_FILE = "conan_{pkgname}_vars{config}.premake5.lua"
+PREMAKE_VAR_FILE = "conan_{pkgname}_vars_{config}.premake5.lua"
```
Is this generating different filenames than before? What happened with the ``_``?
I finally understood that this is because of the changes in ``_config_suffix()``, that don't prepend anymore the first ``_``, but it was initially concerning that this could break, because the filename would be different.

perseoGI replied:
This will not introduce any breaking change but a bug fix in architectures different from `x86_64`, the only supported architecture by the legacy generator.

memsharded commented on conan/tools/premake/premake.py:
```diff
@@ -7,34 +19,99 @@
     @staticmethod
     def _expand_args(args):
-        return ' '.join([f'--{key}={value}' for key, value in args.items()])
+        return " ".join([f"--{key}={value}" for key, value in args.items()])
```
Formatting changes, extra overhead for reviewing

memsharded commented on conan/tools/premake/premake.py:
```diff
@@ -7,34 +19,99 @@
+        if not premake_conan_toolchain.exists():
+            raise ConanException(
+                f"Premake toolchain file does not exist: {premake_conan_toolchain}\nUse PremakeToolchain to generate it."
+            )
```
This will be breaking for all Premake users.
Because their recipes don't have a ``PremakeToolchain`` yet. This can't be done.

perseoGI replied:
Moved this raise to new not-breaking method `build()` and left this check in `configure()` as a distinguish mechanism between legacy and new behavior.

memsharded commented on conan/tools/premake/premake.py:
```diff
@@ -7,34 +19,99 @@
+        """\
+    #!lua
+    include("{{luafile}}")
+    include("{{premake_conan_toolchain}}")
```
This will break existing usage

perseoGI replied:
Latest changes will allow having both legacy behavior and new behavior if `PremakeToolchain` is being used or not

memsharded commented on conan/tools/premake/toolchain.py:
```diff
@@ -0,0 +1,245 @@
+from conan.tools.microsoft.visual import VCVars
+from conan.tools.premake.premakedeps import PREMAKE_ROOT_FILE
+
+_jinja_env = Environment(trim_blocks=True, lstrip_blocks=True)
```
No need for a global here, this can be instantiated in the necessary places, not that important to reuse

memsharded commented on conan/tools/premake/toolchain.py:
```diff
@@ -0,0 +1,245 @@
+        return ", ".join(f'"{item}"' for item in items) if items else None
+
+
+    build_env = self._conanfile.buildenv.vars(self._conanfile)
+    def _get_env_list(env):
+        v = build_env.get(env, [])
+        return v.strip().split() if not isinstance(v, list) else v
+
+    extra_defines = format_list(
+        _get_env_list("DEFINES")
+        + conanfile.conf.get("tools.build:defines", default=[], check_type=list)
+        + self.extra_defines
+    )
+    extra_c_flags = format_list(
+        _get_env_list("CFLAGS")
+        + conanfile.conf.get("tools.build:cflags", default=[], check_type=list)
+        + self.extra_cflags
+    )
+    extra_cxx_flags = format_list(
+        _get_env_list("CXXFLAGS")
+        + conanfile.conf.get("tools.build:cxxflags", default=[], check_type=list)
+        + self.extra_cxxflags
+    )
+    extra_ld_flags = format_list(
```
Is it really necessary to read from Enviroment the CXXFLAGS and others and inject them?
Other generators such as ``CMakeToolchain`` or ``MSBuildToolchain`` don't do it, I am not sure this should be done. I know Meson does it, but it is the exception.

perseoGI replied:
I think it will provide extra flexibility to the users, maybe overkill(?)
It is already tested and works well.
Also, it will provide a similar behavior to `MesonToolchain` as you mentioned.

memsharded replied:
The MesonToolchain is the exception, and it "slipped" there because of legacy reasons, the other generators don't provide it, Meson is the exception, and better not have extra exceptions.

Better not aim for "extra flexibility", but better aim for minimal, user-feedback driven features, otherwise it is too speculative, but adds complexity and maintenance to the code

memsharded commented on test/functional/toolchains/test_premake.py:
```diff
@@ -1,74 +1,124 @@
 
 @pytest.mark.skipif(platform.system() != "Linux", reason="Only for Linux now")
 @pytest.mark.tool("premake")
-def test_premake(matrix_client):
```
Better not change existing tests, unless they are considered broken.
Because these tests protects against breaking existing users.

perseoGI replied:
Okey, restored old test and keep old behaviour when `PremakeToolchain` is not used

memsharded commented on test/integration/toolchains/premake/test_premake.py:
```diff
@@ -1,24 +0,0 @@
-from conan.test.utils.tools import TestClient
-
-
-def test_premake_args():
```
Why remove this test?

perseoGI replied:
Restored

memsharded reviewed (changes requested):
I think this is breaking behavior, please review

perseoGI committed:
Keep both legacy premake generator logic

While introducing new generator by simply enabling PremakeToolchain

perseoGI committed:
Added new tests for premake generators

memsharded commented on conan/tools/premake/toolchain.py:
```diff
@@ -0,0 +1,261 @@
+
+    def _get_env_list(env):
+        v = build_env.get(env, [])
+        return v.strip().split() if not isinstance(v, list) else v
```
Better simplify and drop environment reading.

memsharded commented on test/integration/toolchains/premake/test_premake.py:
```diff
@@ -21,4 +20,82 @@ def build(self):
     c.save({"conanfile.py": conanfile})
     c.run("build . -s compiler=msvc -s compiler.version=193 -s compiler.runtime=dynamic")
-    assert "conanfile.py: Running premake5 --file=myproject.lua vs2022 --myarg=myvalue!!" in c.out
+    print(c.out)
```
```suggestion
```

memsharded commented on test/integration/toolchains/premake/test_premaketoolchain.py:
```diff
@@ -0,0 +1,143 @@
+
+    toolchain = tc.load(
+        os.path.join("build-release", "conan", "conantoolchain.premake5.lua")
+    )
```
```suggestion
    toolchain = tc.load("build-release/conan/conantoolchain.premake5.lua")
```
Keep it simple, this works in all platforms

memsharded reviewed (commented):
Getting closer :)

Assigning it to next 2.18

perseoGI committed:
Drop support for buildenv flags reading

perseoGI committed:
Merge branch 'develop2' into feature/premake_toolchain

perseoGI committed:
Give wasm support to premake toolchain

perseoGI committed:
Improved premake toolchain:

- Support fPIC option
- Support emcc compiler and other architectures
- Added premake exe and lib templates
- Improved test cases detecting new issues
- Fixed arch bug

perseoGI committed:
Added shared option support on premake libs

perseoGI committed:
Fix premake integration tests

perseoGI committed:
Fix tests

perseoGI committed:
Changed dylib for so as test runs in linux

perseoGI committed:
Run premake test only in linux

memsharded commented on pyproject.toml:
```diff
@@ -1,3 +1,6 @@
 [build-system]
 requires = ["setuptools", "wheel"]
 build-backend = "setuptools.build_meta"
+
+
+
```
```suggestion
```

perseoGI committed:
Update pyproject.toml

Co-authored-by: James <memsharded@gmail.com>

perseoGI committed:
Merge branch 'develop2' into feature/premake_toolchain

perseoGI committed:
Added architecture linker specific flags to premake
